README.md
# PS-DueSummaryGem-RubyGems-
PS Due Summary Gem for Packaging Ruby Code with RubyGems 


due_summary-1.0.1/CHANGELOG
Version 1.0.1: The initial due summary is released.
Version 1.0.2: Moved text to duetext.rb to allow for screen text translation.
Version 1.1.0: Added new method to print due summary as a list in addition to table.
Version 2.0.0: Change sort order of measures, and return nil of date is not valid.
Version 2.1.0: New methods to sort and color the measures in different colors
Version 2.1.1: Fixed a bug in the print summary implementation.

duesummary-1.0.0/test/testDue.rb
require 'minitest/autorun'
require_relative '../lib/due'

class TestDue < Minitest::Test

  def test_due_difference_seconds_between_dates
    due = Due::Due.new('2020-10-11 12:00:00')
    due.now_date_time = Time.parse('2020-10-12 12:00:00')
    due.calc_difference

    assert_equal(-86_400.0, due.difference)
  end

  def test_due_difference_minutes_between_dates
    due = Due::Due.new('2020-10-11 12:00:00')
    due.now_date_time = Time.parse('2020-10-12 12:00:00')
    due.calc_difference

    assert_equal(-1440, due.calc_mins_till)
  end

  def test_due_difference_hours_between_dates
    due = Due::Due.new('2020-10-11 12:00:00')
    due.now_date_time = Time.parse('2020-10-12 12:00:00')
    due.calc_difference

    assert_equal(-24, due.calc_hours_till)
  end

  def test_due_difference_days_between_dates
    due = Due::Due.new('2020-10-11 12:00:00')
    due.now_date_time = Time.parse('2020-10-12 12:00:00')
    due.calc_difference

    assert_equal(-1, due.calc_days_till)
  end

  def test_due_difference_days_between_dates
    due = Due::Due.new('2020-10-11 12:00:00')
    due.now_date_time = Time.parse('2020-10-12 12:00:00')
    due.calc_difference

    assert_equal(-1, due.calc_days_till)
  end

  def test_due_difference_months_between_dates
    due = Due::Due.new('2020-07-01 00:00:00')
    due.now_date_time = Time.parse('2020-08-01 00:00:00')
    due.calc_difference

    assert_equal(-1, due.calc_months_till)
  end

  def test_due_difference_years_between_dates
    due = Due::Due.new('2019-07-01 00:00:00')
    due.now_date_time = Time.parse('2020-07-02 00:00:00')
    due.calc_difference

    assert(due.calc_years_till >= -0.99)
  end

  def test_due_difference_weeks_between_dates
    due = Due::Due.new('2020-07-01 00:00:00')
    due.now_date_time = Time.parse('2020-08-01 00:00:00')
    due.calc_difference

    assert(-4.43, due.calc_weeks_till)
  end
end


due_summary-1.0.0/test/testduesummary.rb
require 'minitest/autorun'
require_relative '../lib/due_summary'

class TestDue < Minitest::Test

  def test_due_difference_seconds_between_dates
    due = Due::DueSummary.new('2020-10-11 12:00:00')
    due.now_date_time = Time.parse('2020-10-12 12:00:00')
    due.calc_difference

    assert_equal(-86_400.0, due.difference)
  end

  def test_due_difference_minutes_between_dates
    due = Due::DueSummary.new('2020-10-11 12:00:00')
    due.now_date_time = Time.parse('2020-10-12 12:00:00')
    due.calc_difference

    assert_equal(-1440, due.calc_mins_till)
  end

  def test_due_difference_hours_between_dates
    due = Due::DueSummary.new('2020-10-11 12:00:00')
    due.now_date_time = Time.parse('2020-10-12 12:00:00')
    due.calc_difference

    assert_equal(-24, due.calc_hours_till)
  end

  def test_due_difference_days_between_dates
    due = Due::DueSummary.new('2020-10-11 12:00:00')
    due.now_date_time = Time.parse('2020-10-12 12:00:00')
    due.calc_difference

    assert_equal(-1, due.calc_days_till)
  end

  def test_due_difference_months_between_dates
    due = Due::DueSummary.new('2020-07-01 00:00:00')
    due.now_date_time = Time.parse('2020-08-01 00:00:00')
    due.calc_difference

    assert_equal(-1, due.calc_months_till)
  end

  def test_due_difference_years_between_dates
    due = Due::DueSummary.new('2019-07-01 00:00:00')
    due.now_date_time = Time.parse('2020-07-02 00:00:00')
    due.calc_difference

    assert(due.calc_years_till >= -0.99)
  end

  def test_due_difference_weeks_between_dates
    due = Due::DueSummary.new('2020-07-01 00:00:00')
    due.now_date_time = Time.parse('2020-08-01 00:00:00')
    due.calc_difference

    assert(-4.43, due.calc_weeks_till)
  end
end


due_summary-1.0.1/test/testduesummary.rb
require 'minitest/autorun'
require_relative '../lib/due_summary'

class TestDue < Minitest::Test

  def test_due_difference_seconds_between_dates
    due = Due::DueSummary.new('2020-10-11 12:00:00')
    due.now_date_time = Time.parse('2020-10-12 12:00:00')
    due.calc_difference

    assert_equal(-86_400.0, due.difference)
  end

  def test_due_difference_minutes_between_dates
    due = Due::DueSummary.new('2020-10-11 12:00:00')
    due.now_date_time = Time.parse('2020-10-12 12:00:00')
    due.calc_difference

    assert_equal(-1440, due.calc_mins_till)
  end

  def test_due_difference_hours_between_dates
    due = Due::DueSummary.new('2020-10-11 12:00:00')
    due.now_date_time = Time.parse('2020-10-12 12:00:00')
    due.calc_difference

    assert_equal(-24, due.calc_hours_till)
  end

  def test_due_difference_days_between_dates
    due = Due::DueSummary.new('2020-10-11 12:00:00')
    due.now_date_time = Time.parse('2020-10-12 12:00:00')
    due.calc_difference

    assert_equal(-1, due.calc_days_till)
  end

  def test_due_difference_days_between_dates
    due = Due::DueSummary.new('2020-10-11 12:00:00')
    due.now_date_time = Time.parse('2020-10-12 12:00:00')
    due.calc_difference

    assert_equal(-1, due.calc_days_till)
  end

  def test_due_difference_months_between_dates
    due = Due::DueSummary.new('2020-07-01 00:00:00')
    due.now_date_time = Time.parse('2020-08-01 00:00:00')
    due.calc_difference

    assert_equal(-1, due.calc_months_till)
  end

  def test_due_difference_years_between_dates
    due = Due::DueSummary.new('2019-07-01 00:00:00')
    due.now_date_time = Time.parse('2020-07-02 00:00:00')
    due.calc_difference

    assert(due.calc_years_till >= -0.99)
  end

  def test_due_difference_weeks_between_dates
    due = Due::DueSummary.new('2020-07-01 00:00:00')
    due.now_date_time = Time.parse('2020-08-01 00:00:00')
    due.calc_difference

    assert(-4.43, due.calc_weeks_till)
  end
end


duesummary-1.0.0/lib/due.rb

require 'time'
require 'tabulate'
require_relative 'due/dueText'

# The due module contains all classes and data to provide due-summary functionality.
module Due
  # The due class takes a given date and provides several period measures until or since that date.
  class Due

    # Date time field set via the constructor, and the date-time value the due summary stats are based around.
    attr_accessor :date_time
    # Date time field used to store the current date-time used as part of the due summary calculations.
    attr_accessor :now_date_time
    # The numeric field used to store the time difference between the provided date and the current date and time.
    attr_accessor :difference

    # Class constructor that takes in a DateTime argument to base the due-summary stats on.
    def initialize(the_date_time)
      @date_time = Time.parse(the_date_time)
      @difference = 0.0
    end

    # Uses the puts command to output all the due-summary stats as a line separated text.
    def print_summary
      update_now_date_time
      calc_difference
      puts "#{DueText.minutes} #{till_or_since}: \t #{calc_mins_till.round(2)} #{DueText.minutes_suffix} \n
      #{DueText.hours} #{till_or_since}: \t #{calc_hours_till.round(2)} #{DueText.hours_suffix} \n
      #{DueText.days} #{till_or_since}: \t #{calc_days_till.round(2)} #{DueText.days_suffix} \n
      #{DueText.weeks} #{till_or_since}: \t #{calc_weeks_till.round(2)} #{DueText.weeks_suffix} \n
      #{DueText.months} #{till_or_since}: \t #{calc_months_till.round(2)} #{DueText.months_suffix} \n
      #{DueText.years} #{till_or_since}: \t #{calc_years_till.round(2)} #{DueText.years_suffix}"
    end

    # This method returns an array containing the due-summary stats.
    def summary_array
      row = []
      column = [DueText.minutes, till_or_since, calc_mins_till.round(2), DueText.minutes_suffix]
      row << column
      column = [DueText.hours, till_or_since, calc_hours_till.round(2), DueText.hours_suffix]
      row << column
      column = [DueText.days, till_or_since, calc_days_till.round(2), DueText.days_suffix]
      row << column
      column = [DueText.weeks, till_or_since, calc_weeks_till.round(2), DueText.weeks_suffix]
      row << column
      column = [DueText.months, till_or_since, calc_months_till.round(2), DueText.months_suffix]
      row << column
      column = [DueText.years, till_or_since, calc_years_till.round(2), DueText.years_suffix]
      row << column
    end

    # This method uses the puts command to output all the due-summary stats as a formatted table.
    def summary_table
      update_now_date_time
      calc_difference
      labels = [DueText.period, ' ', DueText.duration, DueText.measure]
      rows = summary_array
      puts tabulate(labels, rows, { 'indent' => 4, 'style' => 'fancy' })
    end

    # This method updates the current date and time regarded as the now current date and time for the due-summary stats.
    def update_now_date_time
      @now_date_time = Time.new
    end

    # This method calculates the due summary time difference between the provided date and the now date.
    def calc_difference
      @difference = @date_time.to_f - @now_date_time.to_f
    end

    # This method returns 'since' or 'till' as a description to describe the time difference.
    def till_or_since
      if @difference.negative?
        DueText.since
      else
        DueText.till
      end
    end

    # This method calculates the due-summary time difference in minutes.
    def calc_mins_till
      (@difference.round(2) / 60.00).round(2)
    end

    # This method calculates the due-summary time difference in hours.
    def calc_hours_till
      (calc_mins_till / 60.00).round(2)
    end

    # This method calculates the due-summary time difference in days.
    def calc_days_till
      (calc_hours_till / 24.00).round(2)
    end

    # This method calculates the due-summary time difference in months.
    def calc_months_till
      (calc_days_till / 31.0).round(2)
    end

    # This method calculates the due-summary time difference in years.
    def calc_years_till
      (calc_months_till / 12.00).round(2)
    end

    # This method calculates the due-summary time difference in weeks.
    def calc_weeks_till
      (calc_days_till / 7.0).round(2)
    end
  end
end


duesummary-1.0.0/lib/due/dueText.rb
# The due module contains all classes and data to provide due-summary functionality.
module Due
  # The dueText class provides all the text for the due-summary functionality.
  class DueText
    def self.minutes
      'Minutes'
    end

    def self.hours
      'Hours'
    end 

    def self.days
      'Days'
    end 

    def self.weeks
      'Weeks'
    end 

    def self.months
      'Months'
    end 

    def self.years
      'Years'
    end 

    def self.minutes_prefix
      'min(s)'
    end

    def self.hours_prefix
      'hours(s)'
    end 

    def self.days_prefix
      'day(s)'
    end 

    def self.weeks_prefix
      'week(s)'
    end 

    def self.months_prefix
      'month(s)'
    end 

    def self.years_prefix
      'year(s)'
    end

    def self.period
      'Period'
    end

    def self.duration
      'Duration'
    end 

    def self.measure
      'Measure'
    end 

    def self.since
      'since'
    end 

    def self.till
      'till'
    end 
  end
end


due_summary-1.0.1/lib/due_summary/duetext.rb
# The due module contains all classes and data to provide due-summary functionality.
module Due
  # The dueText class provides all the text for the due-summary functionality.
  class DueText
    # Method provides text for minutes measure name.
    def self.minutes
      'Minutes'
    end
    # Method provides text for hours measure name.
    def self.hours
      'Hours'
    end 
    # Method provides text for days measure name.
    def self.days
      'Days'
    end 
    # Method provides text for weeks measure name.
    def self.weeks
      'Weeks'
    end 
    # Method provides text for months measure name.
    def self.months
      'Months'
    end 
    # Method provides text for years measure name.
    def self.years
      'Years'
    end 
    # Method provides text for minutes suffix for a messure.
    def self.minutes_suffix
      'min(s)'
    end
    # Method provides text for hours suffix for a messure.
    def self.hours_suffix
      'hours(s)'
    end 
    # Method provides text for days suffix for a messure.
    def self.days_suffix
      'day(s)'
    end 
    # Method provides text for weeks suffix for a messure.
    def self.weeks_suffix
      'week(s)'
    end 
    # Method provides text for months suffix for a messure.
    def self.months_suffix
      'month(s)'
    end 
    # Method provides text for years suffix for a messure.
    def self.years_suffix
      'year(s)'
    end
    # Method provides text for period.
    def self.period
      'Period'
    end
    # Method provides text for duration.
    def self.duration
      'Duration'
    end 
    # Method provides text for measure.
    def self.measure
      'Measure'
    end 
    # Method provides text for since.
    def self.since
      'since'
    end 
    # Method provides text for till.
    def self.till
      'till'
    end 
  end
end


due_summary-1.0.0/lib/due_summary.rb

require 'time'
require 'tabulate'
require_relative 'due_summary/duetext'

# The due module contains all classes and data to provide due-summary functionality.
module Due
  # The due class takes a given date and provides several period measures until or since that date.
  class DueSummary
    # Date time field set via the constructor, and the date-time value the due summary stats are based around.
    attr_accessor :date_time
    # Date time field used to store the current date-time used as part of the due summary calculations.
    attr_accessor :now_date_time
    # The numeric field used to store the time difference between the provided date and the current date and time.
    attr_accessor :difference

    # Class constructor that takes in a DateTime argument to base the due-summary stats on.
    def initialize(the_date_time)
      @date_time = Time.parse(the_date_time)
      @difference = 0.0
    end

    # Uses the puts command to output all the due-summary stats as a line separated text.
    def print_summary
      update_now_date_time
      calc_difference
      puts "#{DueText.minutes} #{till_or_since}: \t #{calc_mins_till.round(2)} #{DueText.minutes_suffix} \n
      #{DueText.hours} #{till_or_since}: \t #{calc_hours_till.round(2)} #{DueText.hours_suffix} \n
      #{DueText.days} #{till_or_since}: \t #{calc_days_till.round(2)} #{DueText.days_suffix} \n
      #{DueText.weeks} #{till_or_since}: \t #{calc_weeks_till.round(2)} #{DueText.weeks_suffix} \n
      #{DueText.months} #{till_or_since}: \t #{calc_months_till.round(2)} #{DueText.months_suffix} \n
      #{DueText.years} #{till_or_since}: \t #{calc_years_till.round(2)} #{DueText.years_suffix}"
    end

    # This method returns an array containing the due-summary stats.
    def summary_array
      row = []
      column = [DueText.minutes, till_or_since, calc_mins_till.round(2), DueText.minutes_suffix]
      row << column
      column = [DueText.hours, till_or_since, calc_hours_till.round(2), DueText.hours_suffix]
      row << column
      column = [DueText.days, till_or_since, calc_days_till.round(2), DueText.days_suffix]
      row << column
      column = [DueText.weeks, till_or_since, calc_weeks_till.round(2), DueText.weeks_suffix]
      row << column
      column = [DueText.months, till_or_since, calc_months_till.round(2), DueText.months_suffix]
      row << column
      column = [DueText.years, till_or_since, calc_years_till.round(2), DueText.years_suffix]
      row << column
    end

    # This method uses the puts command to output all the due-summary stats as a formatted table.
    def summary_table
      update_now_date_time
      calc_difference
      labels = [DueText.period, ' ', DueText.duration, DueText.measure]
      rows = summary_array
      puts tabulate(labels, rows, { 'indent' => 4, 'style' => 'fancy' })
    end

    # This method updates the current date and time regarded as the now current date and time for the due-summary stats.
    def update_now_date_time
      @now_date_time = Time.new
    end

    # This method calculates the due summary time difference between the provided date and the now date.
    def calc_difference
      @difference = @date_time.to_f - @now_date_time.to_f
    end

    # This method returns 'since' or 'till' as a description to describe the time difference.
    def till_or_since
      if @difference.negative?
        DueText.since
      else
        DueText.till
      end
    end

    # This method calculates the due-summary time difference in minutes.
    def calc_mins_till
      (@difference.round(2) / 60.00).round(2)
    end

    # This method calculates the due-summary time difference in hours.
    def calc_hours_till
      (calc_mins_till / 60.00).round(2)
    end

    # This method calculates the due-summary time difference in days.
    def calc_days_till
      (calc_hours_till / 24.00).round(2)
    end

    # This method calculates the due-summary time difference in months.
    def calc_months_till
      (calc_days_till / 31.0).round(2)
    end

    # This method calculates the due-summary time difference in years.
    def calc_years_till
      (calc_months_till / 12.00).round(2)
    end

    # This method calculates the due-summary time difference in weeks.
    def calc_weeks_till
      (calc_days_till / 7.0).round(2)
    end
  end
end
